jetty-util/src/main/java/org/eclipse/jetty/util/resource/ZipFsResource.java

```diff
@@ -0,0 +1,306 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+public class ZipFsResource extends Resource implements Iterable<String>
+{
+    private static final Logger LOG = Log.getLogger(ZipFsResource.class);
+    /**
+     * For a JVM that support JEP-238 Multi-Release jars, use "runtime" behavior.
+     * On older JVMs that don't have JEP-238 support / behavior (eg: Java 1.8), this
+     * FileSystem Environment variable is not used. (it's ignored)
+     */
+    private static final String MR_MODE_RUNTIME = "runtime";
+
+    private final URI zipFsUri;
+    private final Path zipFsPath;
+    private final Path rootPath;
+    private final Map<String, Object> env;
+    private final FileSystem zipFs;
+
+    public ZipFsResource(Path path) throws IOException
+    {
+        this(path.toUri(), MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(File file) throws IOException
+    {
+        this(file.toURI(), MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(URI uri) throws IOException
+    {
+        this(uri, MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(URI uri, String multiReleaseMode) throws IOException
+    {
+        zipFsUri = toZipFsUri(uri);
+        URI fileUri = URI.create(uri.getSchemeSpecificPart());
+        zipFsPath = new File(fileUri.getSchemeSpecificPart()).toPath();
+
+        // The FileSystem Environment
+        env = new HashMap<>();
+        env.put("multi-release", multiReleaseMode);
+
+        // The ZipFs FileSystem
+        FileSystem fs;
+        try
+        {
+            fs = FileSystems.getFileSystem(zipFsUri);
+        }
+        catch (FileSystemNotFoundException ignore)
+        {
+            fs = null;
+        }
+
+        if ((fs == null) || !fs.isOpen())
+        {
+            fs = FileSystems.newFileSystem(zipFsUri, env);
+        }
+
+        zipFs = fs;
+
+        // We only care about the 1 root that a zip/jar has
+        Iterable<Path> rootDirs = zipFs.getRootDirectories();
+        rootPath = rootDirs.iterator().next();
+    }
+
+    private final URI toZipFsUri(URI uri)
+    {
+        String asciiUri = uri.toASCIIString();
+
+        if (asciiUri.equalsIgnoreCase("jar:file:") &&
+                asciiUri.endsWith(".jar"))
+        {
+            // we have what we need already.
+            return uri;
+        }
+
+        if ("file".equalsIgnoreCase(uri.getScheme()))
+        {
+            if (asciiUri.endsWith(".jar"))
+            {
+                return URI.create("jar:" + asciiUri);
+            }
+        }
+
+        throw new UnsupportedOperationException(ZipFsResource.class.getSimpleName() + " only supports [jar:file:/] or [file:/] URIs that end with [.jar] - " + uri);
+    }
+
+    @Override
+    public boolean isContainedIn(Resource resource) throws MalformedURLException
+    {
+        if (resource instanceof PathResource)
```

Comment:
Is this right?  How can a `PathResource` be contained in a `ZipFsResource`?
Surely only a `ZipFsResource` can be contained and it has to have the same `zipFs` and its path within that must be contained within the path of this `ZipFsResource`

Reply:
ZipFsResource represents the JAR itself.
That's what is the FileSystem.  The JAR.
Everything produced from it is a Path, hence a PathResource.